Allow custom matchers to use methods in Pretty

jaredbeck opened the pull request:
When writing a `failure_message` for a custom matcher, the methods in `Pretty` are useful.

Maybe we don't need all of them to be public, but some, like `to_sentence` would be nice.

This PR is just to get the conversation started.

myronmarston commented:
I'm open to exposing this, but have a few suggestions:

* I really don't want `name` or `name_to_sentence`, in their current forms publicly exposed.  Relying upon the `@name` instance variable from the host class has always made me nervous and I don't want to encode that as part of the public API.
* The more I think about it, the more convinced I am that this would work just as well, if not better, as a utility module that supports these methods being called directly on it rather than being a mix-in.  The only aspect that relies upon being mixed in is the `name` stuff I mentioned above, that I'd rather we refactor, anyway.
* `Pretty` is OK as an internal module name, but isn't very descriptive as a public module.  I'm thinking we should rename it.  Maybe `EnglishExpressions` or something is better?

JonRowe commented:
> * The more I think about it, the more convinced I am that this would work just as well, if not better, as a utility module that supports these methods being called directly on it rather than being a mix-in.  The only aspect that relies upon being mixed in is the `name` stuff I mentioned above, that I'd rather we refactor, anyway.
> * `Pretty` is OK as an internal module name, but isn't very descriptive as a public module.  I'm thinking we should rename it.  Maybe `EnglishExpressions` or something is better?

I like both these ideas, feel like tackling it @jaredbeck ?

jaredbeck commented:
> I like both these ideas, feel like tackling it @jaredbeck ?

Heck yeah.  Let me give it a try.

jaredbeck commented:
> Pretty is OK as an internal module name, but isn't very descriptive as a public module. I'm thinking we should rename it. Maybe EnglishExpressions or something is better?

`EnglishPhrases` is a bit easier to type, or how about simply `English`?

jaredbeck commented:
OK, after some heckling from rubocop and yard, I got the travis build passing, so I guess this is ready for review.  Please see the detailed commit message for an overview and rationale.

There's only one thing I'm not happy about.   It's easy to use `English` from within, e.g. `BaseMatcher`, but inside `Matchers.define` I had to use the fully-qualified `RSpec::Matchers::English`, as you can see in `define_matcher.feature`.  This feels awkward, and I'm hoping y'all can suggest a better way.

JonRowe commented:
I'd prefer `EnglishPhrases` over `English`

JonRowe commented on Changelog.md:
```diff
@@ -2,6 +2,11 @@
 
+* Allow perfectionists who write custom matchers to improve their
+  `failure_message` by using methods from RSpec::Matchers::Pretty,
+  now renamed to RSpec::Matchers::English.  Before this, the module
```
This is a bit verbose, should wrap constant names in backticks and you have an excess space after a period. How about:

```
* Expose `RSpec::Matchers::EnglishPhrases` to make it easier to write nice failure messages in custom matchers.
```

JonRowe commented on lib/rspec/matchers/hash_formatting.rb:
```diff
@@ -0,0 +1,20 @@
+module RSpec
+  module Matchers
+    # @api private
```
If these are still to be private they probably could have been left in `Pretty`...

jaredbeck commented:
I've changed the name to `EnglishPhrases` and replaced the changelog entry.

> If these are still to be private they probably could have been left in Pretty...

I think there's some value in separating `Naming` from `HashFormatting`.  For example, there are two tests that include `HashFormatting`, but they don't need to include `Naming`.

JonRowe commented:
> I think there's some value in separating Naming from HashFormatting. For example, there are two tests that include HashFormatting, but they don't need to include Naming.

We try to reduce the number of files we use (because these things actually do matter when it comes to library load times etc) and as it's a private API that we've had in one place for a while I think it's ok to leave them there.

jaredbeck commented:
> We try to reduce the number of files we use ..

Sounds good.  I've re-combined them.

myronmarston commented on lib/rspec/matchers/english_phrases.rb:
```diff
@@ -0,0 +1,58 @@
+        # ActiveSupport users.  It is similar, but has
+        # fewer features.
+        #
+        def humanize(sym)
```
Is `humanize` what AS calls this?

myronmarston replied:
Honestly, I think I liked the old name of `split_words` better...

myronmarston commented on lib/rspec/matchers/english_phrases.rb:
```diff
@@ -0,0 +1,58 @@
+
+        # Converts the given item to string suitable for use in a
+        # list expression.
+        def list_item(obj)
```
I think I'd prefer this be called `item_description`.  The caller is `list `but there's nothing about this logic that is limited to lists.

myronmarston commented on lib/rspec/matchers/english_phrases.rb:
```diff
@@ -0,0 +1,58 @@
+  module Matchers
+    # Facilitates converting ruby objects to English phrases.
+    module EnglishPhrases
+      class << self
```
I have a preference for using `def self.method_name` over opening the singleton class, and that's what we've favored elsewhere. Mind switching it?

jaredbeck replied:
Without the `class << self`, I don't think the `private` keyword will work.

```ruby
module Works
  class << self
    def z; end
    private
    def y; end
  end
end
Works.z #=> nil
Works.y # NoMethodError: private method `y' called for Z:Module

module DoesntWork
  def self.a; end
  private
  def self.b; end
end
DoesntWork.a #=> nil
DoesntWork.b #=> nil
```

Is there another way to do private module methods?

myronmarston commented on lib/rspec/matchers/english_phrases.rb:
```diff
@@ -0,0 +1,58 @@
+module RSpec
+  module Matchers
+    # Facilitates converting ruby objects to English phrases.
+    module EnglishPhrases
```
I think I slightly prefer `EnlighsPhrasing` over `EnglishPhrases` (hard to explain why, though, and it's a very lightly helped opinion).

myronmarston replied:
I think I figured out what I dislike about `EnglishPhrases`...as a plural, it sounds like it's a list of phrases, whereas `EnglishPhrasing` reads more like it helps you create strings using English phrasing (which is more accurate to what it does).

jaredbeck replied:
Makes sense to me.

myronmarston commented on lib/rspec/matchers/built_in/eq.rb:
```diff
@@ -20,7 +20,8 @@ def failure_message_when_negated
         def description
-          "#{name_to_sentence} #{@expected.inspect}"
+          english_name = EnglishPhrases.humanize(name)
+          "%s %s" % [english_name, @expected.inspect]
```
I much prefer using string interpolation to using format strings and `%` as you've done here and below.  Among other things, this form does an extra array allocation.

jaredbeck replied:
OK, I'll switch back to interpolation.  If performance were not a concern, I like format strings because the non-interpolated content (in this case, whitespace) is easier to see.

myronmarston commented on lib/rspec/matchers/english_phrases.rb:
```diff
@@ -0,0 +1,58 @@
+          else
+            " #{items[0...-1].join(', ')}, and #{items[-1]}"
+          end
+        end
```
I think the leading space is a little odd in this.  I know it's how it worked before but I think it will surprise users.  You mind removing it and updating the call sites?

jaredbeck replied:
Good point.  It would definitely surprise people.

jaredbeck replied:
Hmm.  There isn't always a leading space.  If we remove the leading space here, we could end up duplicating the "whitespace logic" in each of the call sites.

myronmarston commented:
> Is there another way to do private module methods?

Use `private_class_method :method_name`.

myronmarston commented:
> Hmm. There isn't always a leading space. If we remove the leading space here, we could end up duplicating the "whitespace logic" in each of the call sites.

Good point.  I think you should add a `@note` YARD doc tag mentioning that the returned string has a leading space except when given an empty list.  That way we're explicitly telling people about it.

jaredbeck commented:
> Use private_class_method ..
> 
> .. add a `@note` YARD doc tag ..

Done and done.  I also added unit tests for `.split_words` and `.list` (spec/rspec/matchers/english_phrasing_spec.rb) and I look forward to any feedback you might have there.

jaredbeck commented:
>  I also added unit tests ..

If you're satisfied with the unit tests, I'd suggest that they replace the cucumber test I wrote.  The unit tests are faster, and the cucumber test feels more like a test than like documentation.

JonRowe commented:
I feel we should keep the cuke as it helps expose this feature...

myronmarston commented on lib/rspec/matchers/built_in/base_matcher.rb:
```diff
@@ -53,11 +53,12 @@ def match_unless_raises(*exceptions)
-          return name_to_sentence unless defined?(@expected)
-          "#{name_to_sentence}#{to_sentence @expected}"
+          desc = EnglishPhrasing.split_words(name)
+          desc += EnglishPhrasing.list(@expected) if defined?(@expected)
```
For perf reasons we can use `<<` here rather than `+=` and allocate one less string.

myronmarston commented on lib/rspec/matchers/dsl.rb:
```diff
@@ -336,7 +340,7 @@ class Matcher
 
-        # Converts matcher name and expected args to an English expresion.
+        # Facilitates better descriptions and failure messages.
         include RSpec::Matchers::Pretty
```
Can we do away with this module include entirely?

jaredbeck replied:
I don't know.  It doesn't look easy.  `Pretty#name` is used in at least one place: `Matcher#inspect`.  

However, I think we can make `Pretty#underscore` private, and I have done so.

jaredbeck replied:
> However, I think we can make Pretty#underscore private, and I have done so.

On second thought, though `Pretty#underscore` was designated `@api private`, I don't know exactly what that means in this project, so making `#underscore` private may or may not constitute a breaking change.  Please advise.

myronmarston commented on spec/rspec/matchers/english_phrasing_spec.rb:
```diff
@@ -0,0 +1,76 @@
+
+        it "first casts its argument to string" do
+          x = double
+          allow(x).to receive(:to_s).and_return("banana")
```
These two lines can be simplified to one:

``` ruby
double(:to_s => "banana")
```

myronmarston commented on spec/rspec/matchers/english_phrasing_spec.rb:
```diff
@@ -0,0 +1,76 @@
+            banana = Struct.new("Banana", :flavor)
+            allow(banana).to receive(:inspect).and_return("Delicious")
+            expect(described_class.list(banana)).to eq(" Delicious")
+          end
```
`banana` is a struct class, not a struct instance, and it doesn't satisfy the `Struct ===` check of the `list` method:

``` ruby
irb(main):002:0> Person = Struct.new(:name)
=> Person
irb(main):003:0> Struct === Person
=> false
irb(main):004:0> Struct === Person.new
=> true
```

You'd have to do `banana.new` to get an instance that satisfies that check.

It also seems a bit smelly to stub a core method like `inspect`...you could just use `banana_instance.inspect` in the expectation instead.

myronmarston commented on spec/rspec/matchers/english_phrasing_spec.rb:
```diff
@@ -0,0 +1,76 @@
+            let(:list) { [double] }
+            it "returns description, and a leading space" do
+              expect(described_class.list(list)).to eq(" Banana")
+              expect(described_class).to \
```
I'm really not a fan of the use of `\` between `to` and the matcher.  It's not a style we use anywhere else in RSpec's specs. You mind putting this all on one line?  If you feel some of these are two long, you can use the styles we use elsewhere:

``` ruby
expect(
  some_long_expression.that_goes_on_its_own_line
).to something

# or

expect(some_expression).to matcher(
  matcher_args,
  can, go,
  on, other_lines
)
```

(Or some combination thereof).

myronmarston commented:
> I feel we should keep the cuke as it helps expose this feature...

I'm :-1: on this.  IMO, the cukes are not an effective way to document every API.  They are there to document the main use case but I don't expect this to be a commonly used API.  The YARD docs are better for that.  Plus having the cuke has a cost that the YARD docs have, in terms of longer build times, etc.

myronmarston commented:
@jaredbeck -- sorry about the delay.  I left some more comments but this is getting close to ready to merge.

jaredbeck commented:
> @jaredbeck -- sorry about the delay. I left some more comments but this is getting close to ready to merge.

No worries.  Thanks for the great comments, I'm learning from them.  I'll ping again when I've addressed them.  Let me know what you guys decide about the cuke.

jaredbeck commented:
> Thanks for the great comments .. I'll ping again when I've addressed them. 

Ready for another review, thanks!

myronmarston commented on features/custom_matchers/define_matcher.feature:
```diff
@@ -335,3 +335,37 @@ Feature: define matcher
+    And the output should contain all of these:
+      | 2 examples, 0 failures                          |
+      | should descend from Fruit, Food, and Object     |
+      | should not descend from Fruit, Food, and Object |
```
I'm still in favor of this scenario being removed, given the spec coverage you added.

jaredbeck commented:
> I'm still in favor of this [cuke] scenario being removed, given the spec coverage you added.

Removed.

myronmarston commented:
Thanks, @jaredbeck!

jaredbeck commented:
> Thanks, @jaredbeck!

Thank *you*.  It was fun.
